app/src/sm_at_dfu.c

```diff
@@ -0,0 +1,605 @@
+#include <nrf_modem_bootloader.h>
+#include <dfu/dfu_target.h>
+#include <dfu/dfu_target_modem_delta.h>
+#include <dfu/dfu_target_mcuboot.h>
+#include "sm_at_host.h"
+#include "sm_at_dfu.h"
+#include "sm_settings.h"
+#include "sm_uart_handler.h"
+
+LOG_MODULE_REGISTER(sm_dfu, CONFIG_SM_LOG_LEVEL);
+
+#define APP_DFU_BUFFER_SIZE CONFIG_SM_DATAMODE_BUF_SIZE
+
+enum xdfu_image_type {
+	DFU_TYPE_APP = 0,
+	DFU_TYPE_DELTA_MFW = 1,
+	DFU_TYPE_FULL_MFW = 2,
+};
+
+enum xdfu_full_mfw_segment_type {
+	DFU_FULL_MFW_SEGMENT_BOOTLOADER = 0,
+	DFU_FULL_MFW_SEGMENT_FIRMWARE = 1,
+};
+
+enum xdfu_operation {
+	DFU_OPERATION_INITIALIZE = 0,
+	DFU_OPERATION_DATA_WRITE = 1,
+	DFU_OPERATION_APPLY_UPDATE = 2,
+};
+struct xdfu_app_datamode_context {
+	size_t addr;
+	size_t len;
+} xdfu_app_datamode_context;
+
+struct xdfu_delta_mfw_datamode_context {
+	size_t addr;
+	size_t len;
+} xdfu_delta_mfw_datamode_context;
+
+struct xdfu_full_mfw_datamode_context {
+	size_t addr;
+	size_t len;
+} xdfu_full_mfw_datamode_context;
+
+bool sm_bootloader_mode_requested;
+bool sm_bootloader_mode_enabled;
+
+int full_mfw_dfu_segment_type = DFU_FULL_MFW_SEGMENT_BOOTLOADER;
+
+static uint8_t app_dfu_buffer[APP_DFU_BUFFER_SIZE] __aligned(4);
+static bool app_dfu_buffer_initialized;
+
+static enum xdfu_image_type xdfu_current_image_type;
+static uint32_t xdfu_bytes_written;
+static int xdfu_status;
+
+static void delta_dfu_evt_handler(enum dfu_target_evt_id evt_id)
+{
+	switch (evt_id) {
+	case DFU_TARGET_EVT_ERASE_PENDING:
+		LOG_INF("Delta DFU erase pending");
+		break;
+	case DFU_TARGET_EVT_TIMEOUT:
+		LOG_WRN("Delta DFU erase timeout");
+		break;
+	case DFU_TARGET_EVT_ERASE_DONE:
+		LOG_INF("Delta DFU erase done");
+		break;
+	default:
+		break;
+	}
+}
+
+int bootloader_mode_request(bool enable)
+{
+	int err;
+
+	sm_bootloader_mode_requested = enable;
+
+	err = sm_settings_bootloader_mode_save();
+	if (err) {
+		LOG_ERR("Failed to set bootloader mode requested to: %s",
+			enable ? "enabled" : "disabled");
+		return err;
+	}
+
+	LOG_DBG("Bootloader mode request set to: %s", enable ? "enabled" : "disabled");
+
+	return 0;
+}
+
+static int set_full_mfw_dfu_segment_type(enum xdfu_full_mfw_segment_type type)
+{
+	int err;
+
+	full_mfw_dfu_segment_type = type;
+
+	err = sm_settings_full_mfw_dfu_segment_type_save();
+	if (err) {
+		LOG_ERR("Failed to set full MFW DFU segment type to: %d", type);
+		return err;
+	}
+
+	LOG_DBG("Full MFW DFU segment type set to: %d", type);
+
+	return 0;
+}
+
+static int xdfu_datamode_callback(uint8_t op, const uint8_t *data, int len, uint8_t flags)
+{
+	ARG_UNUSED(flags);
+
+	int err;
+
+	switch (op) {
+	case DATAMODE_SEND:
+		if (data == NULL || len <= 0) {
+			LOG_ERR("Chunk data invalid (data=%p len=%d)", (void *)data, len);
+			return -EINVAL;
+		}
+
+		switch (xdfu_current_image_type) {
+		case DFU_TYPE_APP:
+			err = dfu_target_mcuboot_write((const void *)data, len);
+			if (err) {
+				LOG_ERR("Failed to write app firmware: %d", err);
```

Comment:
Could use the following approach to save a bit of flash in the error prints.

```suggestion
				LOG_ERR("Failed to write %s: %d", err, "app firmware");
```